sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala

```diff
@@ -0,0 +1,328 @@
+     * @note The UnsafeRow returned is reused across calls, and the UnsafeRow just points to
+     *       the given byte array.
+     */
+    def decodeKey(keyBytes: Array[Byte]): UnsafeRow = {
+      if (keyBytes != null) {
+        // Platform.BYTE_ARRAY_OFFSET is the recommended way refer to the 1st offset. See Platform.
+        keyRow.pointTo(
+          keyBytes,
+          Platform.BYTE_ARRAY_OFFSET + STATE_ENCODING_NUM_VERSION_BYTES,
+          keyBytes.length - STATE_ENCODING_NUM_VERSION_BYTES)
+        keyRow
+      } else null
+    }
+
+    /**
+     * Decode byte array for a value to a UnsafeRow.
+     *
+     * @note The UnsafeRow returned is reused across calls, and the UnsafeRow just points to
+     *       the given byte array.
+     */
+    def decodeValue(valueBytes: Array[Byte]): UnsafeRow = {
+      if (valueBytes != null) {
+        // Platform.BYTE_ARRAY_OFFSET is the recommended way refer to the 1st offset. See Platform.
+        valueRow.pointTo(
+          valueBytes,
+          Platform.BYTE_ARRAY_OFFSET + STATE_ENCODING_NUM_VERSION_BYTES,
+          valueBytes.size - STATE_ENCODING_NUM_VERSION_BYTES)
+        valueRow
+      } else null
+    }
+
+    /**
+     * Decode pair of key-value byte arrays in a pair of key-value UnsafeRows.
+     *
+     * @note The UnsafeRow returned is reused across calls, and the UnsafeRow just points to
+     *       the given byte array.
+     */
+    def decode(byteArrayTuple: ByteArrayPair): UnsafeRowPair = {
+      rowTuple.withRows(decodeKey(byteArrayTuple.key), decodeValue(byteArrayTuple.value))
+    }
+  }
+}
+
+object RocksDBStateStoreProvider {
+  // Version as a single byte that specifies the encoding of the row data in RocksDB
+  val STATE_ENCODING_NUM_VERSION_BYTES = 1
+  val STATE_ENCODING_VERSION: Byte = 0
+
+  // Native operation latencies report as latency per 1000 calls
+  // as SQLMetrics support ms latency whereas RocksDB reports it in microseconds.
+  val CUSTOM_METRIC_GET_TIME = StateStoreCustomTimingMetric(
+    "rocksdbGetLatency", "RocksDB: avg get latency (per 1000 calls)")
+  val CUSTOM_METRIC_PUT_TIME = StateStoreCustomTimingMetric(
+    "rocksdbPutLatency", "RocksDB: avg put latency (per 1000 calls)")
+
+  // Commit latency detailed breakdown
+  val CUSTOM_METRIC_WRITEBATCH_TIME = StateStoreCustomTimingMetric(
+    "rocksdbCommitWriteBatchLatency", "RocksDB: commit - write batch time")
+  val CUSTOM_METRIC_FLUSH_TIME = StateStoreCustomTimingMetric(
+    "rocksdbCommitFlushLatency", "RocksDB: commit - flush time")
+  val CUSTOM_METRIC_PAUSE_TIME = StateStoreCustomTimingMetric(
+    "rocksdbCommitPauseLatency", "RocksDB: commit - pause bg time")
+  val CUSTOM_METRIC_CHECKPOINT_TIME = StateStoreCustomTimingMetric(
+    "rocksdbCommitCheckpointLatency", "RocksDB: commit - checkpoint time")
+  val CUSTOM_METRIC_FILESYNC_TIME = StateStoreCustomTimingMetric(
+    "rocksdbCommitPauseBgTime", "RocksDB: commit - file sync time")
```

Comment:
rocksdbCommitPauseBgTime? Or FileSyncTime?

Reply:
Yea, changed to rocksdbFileSyncTime